proposal/interleaved-declarations.md

```diff
@@ -0,0 +1,125 @@
+# Interleaved Declarations: Draft 1
+
+*([Issue](https://github.com/sass/sass/issues/3846))*
+
+## Table of Contents
+
+* [Background](#background)
+* [Summary](#summary)
+* [Definitions](#definitions)
+  * [Current Style Rule](#current-style-rule)
+* [Declarations](#declarations)
+  * [Semantics](#semantics)
+* [Unknown At-Rules](#unknown-at-rules)
+  * [Semantics](#semantics-1)
+* [Deprecation Process](#deprecation-process)
+
+## Background
+
+> This section is non-normative.
+
+Historically, Sass has hoisted declarations that appear after nested rules into
+the original style rule. Although this doesn't match the logical ordering of the
+source document, it *does* avoid duplicating the style rule (and any other
+nested rules) for each set of interleaved properties.
+
+Many years after this decision was made, the [CSS Nesting Module] was published
+by the W3C, adding native support for nesting rules within style rules using
+essentially the same syntax as Sass. Although initially its semantics for
+interleaved declarations were the same as Sass's, [a later update] (four days
+ago at time of writing) changed it to preserve the logical ordering of
+interleaved declarations.
+
+[CSS Nesting Module]: https://drafts.csswg.org/css-nesting/
+[a later update]: https://github.com/w3c/csswg-drafts/commit/e5547b96f5de6fb5a68d050f42d562c448b99d0b
+
+## Summary
+
+> This section is non-normative.
+
+This proposal changes the behavior of declarations that appear after nested
+rules. Instead of being hoisted to the beginning of the parent rule, they will
+now be placed in their original logical order, duplicating the parent selector.
+
+That is, the following Sass:
+
+```scss
+.a {
+  color: red;
+  @media screen {color: blue}
+  color: green;
+}
+```
+
+will now compile to this CSS:
+
+```scss
+.a {
+  color: red;
+}
+
+@media screen {
+  .a {
+    color: blue;
+  }
+}
+
+.a {
+  color: green;
+}
+```
+
+## Definitions
+
+### Current Style Rule
+
+Change the definition of the [current style rule][old] to:
+
+[old]: ../spec/style-rules.md#current-style-rule
+
+> Differences are highlighted in bold.
+
+The *current style rule* is the CSS style rule that was created by the innermost
+[execution of a style rule], `@media` rule, `@supports` rule, unknown at-rule,
+**or [declaration]**.
+
+[execution of a style rule]: ../spec/style-rules.md#semantics
+[declaration]: #semantics
+
+## Declarations
+
+### Semantics
+
+Add the following to the semantics for executing a declaration `declaration`
+before "Append `css` to `parent`":
+
+* If `parent` isn't the last statement in its parent:
+
+  * Let `copy` by a copy of `parent` without any children.
+
+  * Append `copy` to `parent`'s parent.
+
+  * Set `parent` to `copy`.
```

Comment:
I'm slightly confused by this part. From the example above:
```scss
.a {
  color: red;
  @media screen {color: blue}
  color: green;
}
```

1. `copy` = `@media screen{}`
2. `copy` (`@media screen{}`) is appended to `parent`'s parent (`.a{color: red}`)
3. `parent` = `copy`

This would mean (IIUC) that it would produce the same(?) as the input:

```css
.a {
  color: red;
  @media screen {color: blue}
  color: green;
}
```

Or at the very least, with this wording I don't get that `@media` will wrap `.a`. 
I feel like it's missing "make an empty copy of all ancestors, and insert them inside `copy`, then make the innermost element inside `copy` the new `parent`" or something like that.

Or is that implied given how the other rules are executed?

Reply:
When you're executing `color: blue`, the structure of the generated CSS will be:

```css
a {color: red}
@media screen {a {}}
```

and `parent` will be `a {}`. This _is_ the last statement in its parent, so the new condition doesn't trigger.

Where it matters is when you're executing `color: green`. At this point, the CSS will be:

```css
a {color: red}
@media screen {a {color: blue}}
```

and `parent` is `a {color: red}`. This is _not_ the last statement in its parent, because the `@media` rule comes after it. So:

1. `copy` = `a {}`
2. `copy` (`a {}`) is appended to `parent`'s parent (the root stylesheet)
3. `parent` = `copy`

Then `color: green` is appended to `parent` and you get:

```css
a {color: red}
@media screen {a {color: blue}}
a {color: green}
```

This _does_ make me realize though that the current spec implies that the spec as written implies that the parent selector is duplicated anew for _each_ interleaved declaration, which isn't right.

Reply:
gotcha, that makes sense